### ISSUE TITLE
Blazor WASM: File Upload - MultipartFormDataContent.Add() throws internal Invalid JSON exception

### ISSUE THREAD

**Issue by Grizzlly** [dotnet/aspnetcore#39087]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Uploading a file results in `Invalid JSON` in internal code.

`ImageUploader.razor`:
```razor
<InputFile OnChange="UploadFiles" multiple  />
```

`ImageUploader.razor.cs`:
```c#
using BlazorTests.Shared;
using Microsoft.AspNetCore.Components;
using Microsoft.AspNetCore.Components.Forms;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;

namespace BlazorTests.Client.Shared
{
    public partial class ImageUploader : ComponentBase
    {
        private List<File> files = new();
        private List<UploadResult> uploadResults = new();
        private int maxAllowedFiles = 10;

        private bool shouldRender;

        protected override bool ShouldRender() => shouldRender;

        private void UploadFiles(InputFileChangeEventArgs e)
        {
            shouldRender = false;
            long maxFileSize = 1024 * 1024 * 15;
            var upload = false;

            using var content = new MultipartFormDataContent();


            foreach (var file in e.GetMultipleFiles(maxAllowedFiles))
            {
                if (uploadResults.SingleOrDefault(
                    f => f.FileName == file.Name) is null)
                {
                    try
                    {
                        var fileContent =
                            new StreamContent(file.OpenReadStream(maxFileSize));

                        fileContent.Headers.ContentType =
                            new MediaTypeHeaderValue(file.ContentType);

                        files.Add(new() { Name = file.Name });
                        try // This is the block that throws. Note that the exception is not catched and the app breaks.
                        {
                            content.Add(
                            content: fileContent,
                            name: "\"files\"",
                            fileName: file.Name);
                        }catch(JsonException ex)
                        {
                            Console.WriteLine(ex.Message);
                        }

                        upload = true;
                    }
                    catch (Exception)
                    {
                        uploadResults.Add(
                            new()
                            {
                                FileName = file.Name,
                                ErrorCode = 6,
                                Uploaded = false
                            });
                    }
                }
            }

            shouldRender = true;
        }

        private class File
        {
            public string Name { get; set; }
        }
    }
}

```

The code is taken from [documentation](https://docs.microsoft.com/en-us/aspnet/core/blazor/file-uploads?view=aspnetcore-6.0&pivots=webassembly#upload-files-to-a-server).

### Expected Behavior

No exception, obviously.

### Steps To Reproduce

https://github.com/Grizzlly/BlazorTests

### Exceptions (if any)

When I debug and try to upload a file, I get this:
![image](https://user-images.githubusercontent.com/19261573/146458973-7cae7c4e-6602-453d-af25-43354f129ad9.png)

### .NET Version

6.0.100

### Anything else?

.NET SDK (reflecting any global.json):
 Version:   6.0.100
 Commit:    9e8b04bbff

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22000
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100\

Host (useful for support):
  Version: 6.0.0
  Commit:  4822e3c3aa

.NET SDKs installed:
  3.1.415 [C:\Program Files\dotnet\sdk]
  5.0.104 [C:\Program Files\dotnet\sdk]
  5.0.209 [C:\Program Files\dotnet\sdk]
  5.0.303 [C:\Program Files\dotnet\sdk]
  5.0.403 [C:\Program Files\dotnet\sdk]
  6.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-rc.1.21451.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-rc.1.21451.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download

**Comment by Grizzlly** [comment]
Update: This might be a recent issue as file uploading used to work about 1 week ago and since then I have updated Visual Studio from the installer.

**Comment by TanayParikh** [comment]
Thanks @Grizzlly for the report. @guardrex have you had reports on this not working. I just copied the [sample code](https://docs.microsoft.com/en-us/aspnet/core/blazor/file-uploads?view=aspnetcore-6.0&pivots=webassembly#upload-component) from the upload component and put it into a clean `blazorwasm` app and I get this exception.

**Comment by guardrex** [comment]
I haven't had such a report.

Side Note (that should really matter perhaps): We don't implement as a partial class. @Grizzlly, did you try the code in a component as shown in the doc to see if the same error occurs?

**Comment by TanayParikh** [comment]
> did you try the code in a component in the **`Client`** app as shown in the doc to see if the same error occurs

Looks like it: https://github.com/Grizzlly/BlazorTests/blob/master/Client/Shared/ImageUploader.razor.cs

**Comment by guardrex** [comment]
Sorry that I didn't take a closer look. I'm 😩 and OOF until Monday, really. We don't show it implemented as a partial class. I can check and test a bit and see if I can repro this, but it wouldn't be until Monday.

**Comment by TanayParikh** [comment]
I wonder if this is related to https://github.com/dotnet/aspnetcore/issues/38962. However that's been around since .NET 5, which would imply this has been broken since then (which I find hard to believe). 

> Update: This might be a recent issue as file uploading used to work about 1 week ago and since then I have updated Visual Studio from the installer.

We only released the `6.0.101` patch last week, however you're running on `6.0.100` so that shouldn't have changed anything here. VS updates also don't play a part in this type of issue.

**Comment by TanayParikh** [comment]
> I'm 😩 and OOF until Monday, really.

Oh didn't know, no worries at all, no rush here.

**Comment by guardrex** [comment]
No worries. I'm just out of hours for the week ...... and toast ..... 😄 ..... for the **_YEAR!_** 😆 LOL

I'll take a look first thing Monday morning.

**Comment by Grizzlly** [comment]
Updated to `6.0.101` and the problem persists. Also tried moving the code from the partial class to `@code` block but I get the same error.

> VS updates also don't play a part in this type of issue.
Sorry, I was thinking about the `WebAssembly` NuGet package...

@guardrex @TanayParikh Apparently, downgrading the `WebAssembly` package to `6.0.0-preview.6.21355.2` appears to be working as it does not throw. For me, everything from preview 7 onwards, including 6.0.1, throws `Invalid JSON`. This is weird because it was working one week ago... Maybe it is a browser issue?

I have added `edge://version` output to the repo in `browser-version.txt`

As a side question,
> We don't implement as a partial class.

Why would this make a difference? I have always used the code-behind approach and did not think that it might cause problems...

**Comment by guardrex** [comment]
It doesn't now that you've tested. I just want to make sure that it's ruled out, including a conversion problem going from the doc's example to the partial. I still plan to take a look and see if I can repro on Monday.

**Comment by Grizzlly** [comment]
> I wonder if this is related to #38962

I see that you have made and merged a PR that closed that issue. Might be a good idea to test with that... 

Unfortunately, I could not get the nightly build to run :(

**Comment by Grizzlly** [comment]
Changing the `content` parameter to

`content: new StringContent(JsonSerializer.Serialize(fileContent), Encoding.UTF8, "application/json")`

or

`content: new StringContent(await fileContent.ReadAsStringAsync(), Encoding.UTF8, "application/octet-stream")`

seems to be working. I feel like anything with `StringContent` works, but I have not done further testing. Maybe the problem is with the `StreamContent` class?

**Comment by guardrex** [comment]
I can't repro a problem with the doc's example (as provided, i.e., not using a partial class) ...

Upload two test files, a text file and a PNG image ...

<img width="726" alt="Capture" src="https://user-images.githubusercontent.com/1622880/146770835-f2e52de2-1167-412c-a16e-d67645712dda.PNG">

Files are uploaded ...

<img width="229" alt="Capture1" src="https://user-images.githubusercontent.com/1622880/146770870-420dbadd-1b86-4c0e-ac5c-70fad55dd35b.PNG">

Change the text file's extension and open it ...

<img width="407" alt="Capture2" src="https://user-images.githubusercontent.com/1622880/146770883-62ec2c00-ea79-473d-ad74-c84be035caa4.PNG">

Change the image file's extension and open it ...

<img width="436" alt="Capture3" src="https://user-images.githubusercontent.com/1622880/146770909-94fea7e4-9f5f-4e6c-ad4b-792abfbc63d3.PNG">

I made a couple of nit updates this morning to the topic WRT namespaces, nothing that has any bearing on this 🐛 report. I'll keep an 👁️ on the issue here for further investigation to see what you discover for possible doc/example updates later.

@Grizzlly ... I know it sounds like a strange request here ... but if you didn't verify the code breaking with the error you found **_implementing the example exactly as the topic shows_** (i.e., don't use a partial), I recommend that you create a new test app and confirm it breaks as shown. I just like to rule out, for example, that funky namespaces aren't the cause. I'd be happy to test further (i.e., use a partial), but there will be a delay. I'm *on vacation* ⛄, and I really, **_REALLY_** need to get a break and get off of the repo for a bit. 🏃😩😄 If you don't want to run the test tho, I'll continue to test (i.e., with a partial), but it will be when I get back from vacation in '22. My main reason for working the doc's example (**_on vacation_** 😩) is that I hate *gremlins in the docs* 😈 so much that I'll immediately fix any known doc bug on the spot ... 24/7/365 ... vacation or not. Anyway ... I can't detect a problem thus far with the topic's guidance, so it isn't actionable for docs ... **_yet_** 😄.

**Comment by Grizzlly** [comment]
Created a [new project](https://github.com/Grizzlly/NewFileTest), .NET 6.0.101

Creation options:

- HTTPS
- ASP.NET Core Hosted
- Progressive Web Application

![image](https://user-images.githubusercontent.com/19261573/146781867-b059f23c-f833-485c-a684-2adc86fcf410.png)

Again, I could track the exception down to `content.Add()`, specifically `fileContent`. If I change `fileContent` to a `StringContent` object, it works.

Both this and #38962 show a problem with `EndInvokeJS`.

**Comment by guardrex** [comment]
Ah ... sorry if I missed that it's a hosted PWA. You may have called that out earlier. We state ... *because Steve says so* 😄 ... that the hosting and deployment model aren't relevant to PWAs; however, network behaviors (and possibly JS interop behaviors) aren't necessarily the same. How about a test without the PWA switch (i.e., pure hosted WASM)? I just want to try and narrow down the possibilities here.

Although offline support isn't enabled with a PWA while in the Development environment, idk what the JS interop ramifications are for such an app under these conditions. One thing that you might try with that app (the PWA version) is to go ahead and publish it ... then study this 🐛 when offline mode is enabled. It's another angle because Steve remarks that online/offline network behaviors can be different. Your app might work **_when published_**.

**Comment by Grizzlly** [comment]
Update: Even with PWA support disabled, I get the same exception. It also throws when published. The project can be found [here](https://github.com/Grizzlly/NewFileTestNoPWA).

**Comment by guardrex** [comment]
Ok ... thanks for checking. I can't repro the behavior (yet), so we'll need to wait until the PU gets back on this. I'll eventually perform a hosted PWA test here. I'll be back the first week of '22, and I'll get to it *ASAP* depending on any priority issues that must be addressed first when I get back.

**Comment by Grizzlly** [comment]
Thank you! It is a bummer that this broke out of the blue as my app kinda depends on this... Let me know if you need any more info.

**Comment by RobertBouillon** [comment]
I'm also encountering this issue. Not using PWA, but am using pure WASM. Just upgraded up from .NET 5.

**Comment by guardrex** [comment]
@TanayParikh ... I can't repro a problem for pure hosted WASM for the doc guidance. I haven't tried for hosted PWA yet. I was going to try that after I return from vacation.

**Comment by TanayParikh** [comment]
I got the issue to repro using a clean blazorwasm app with the documented upload component [here](https://docs.microsoft.com/en-us/aspnet/core/blazor/file-uploads?view=aspnetcore-6.0&pivots=webassembly#upload-component). This may end up requiring some doc changes, but at the core I believe this may be a product issue. I'm investigating accordingly 😄 



> I'm also encountering this issue. Not using PWA, but am using pure WASM. Just upgraded up from .NET 5.

In the meantime, you can use @Grizzlly's recommendation of `StringContent` instead of `StreamContent` [here](https://github.com/dotnet/aspnetcore/issues/39087#issuecomment-997174943)

**Comment by TanayParikh** [comment]
Thanks all for the discussion, I've confirmed this is a duplicate of https://github.com/dotnet/aspnetcore/issues/38962 and is fixed by https://github.com/dotnet/aspnetcore/pull/39060.

Given this is currently in a broken state, I believe taking @Grizzlly's suggestion [from here](https://github.com/dotnet/aspnetcore/issues/39087#issuecomment-997174943) and using the octet-stream should be the appropriate workaround. @guardrex when you're back from vacation do you mind updating the docs?

**Comment by guardrex** [comment]
Yes, I'll open an issue and take care of it.

**Comment by TanayParikh** [comment]
https://github.com/dotnet/aspnetcore/pull/39060 will ensure we don't get the un-necessary exceptions, however the fact that we threw an exception in the first place in .NET 6 and not .NET 5 was bothering me. I wanted to make sure this wasn't indicative of a deeper issue. Fortunately it's not.

- We're reading a stream, and before processing the stream request in .NET we're cancelling the request.
  - This leads to the incoming call to `EndInvokeJS` failing as the pending task is missing. We throw an exception saying we haven't read the contents of the call. This is what https://github.com/dotnet/aspnetcore/pull/39060 fixes.
- In .NET 6 we introduced streaming, and the streaming mechanism is what's sending the initial "prep" request that's being cancelled. This is why we don't get this error in .NET 5. 
  - Specifically, in .NET we call `e.File.OpenReadStream()` which requests the stream, and JS sends back the `__jsStreamReferenceLength` and `__jsObjectId`. However in between the request and response, we cancel (ex. end of function or a variety of other reasons). 



I've minimized the repro to the following. 

```razor
@page "/"
@using Microsoft.AspNetCore.Components.Forms
@using System.Linq
@using System.Net.Http
@using System.Net.Http.Headers
@using Microsoft.Extensions.Logging
@inject HttpClient Http

<PageTitle>Index</PageTitle>


<p>
    <label>
        <InputFile OnChange="@OnInputFileChange" multiple />
    </label>
</p>

@code {
    private void OnInputFileChange(InputFileChangeEventArgs e)
    {
        using var content = new MultipartFormDataContent();

        var fileContent = new StreamContent(e.File.OpenReadStream());

        content.Add(fileContent);

        // await Task.Delay(10_000); // Uncomment to stop exception (as we no longer cancel before the response is interpreted)
        Console.WriteLine("Hi I'm here without fail");
    }
}
```


<details>
<summary>Exceptions from investigation</summary>

Stack trace on what's removing the pending task:

```
       Uncaught (in promise) Error: System.Exception:    at System.Environment.get_StackTrace()
   at Microsoft.JSInterop.JSRuntime.CleanupTasksAndRegistrations(Int64 taskId) in /workspaces/aspnetcore/src/JSInterop/Microsoft.JSInterop/src/JSRuntime.cs:line 144
   at Microsoft.JSInterop.JSRuntime.<>c__DisplayClass18_0`1[[Microsoft.JSInterop.IJSStreamReference, Microsoft.JSInterop, Version=7.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]].<InvokeAsync>b__0() in /workspaces/aspnetcore/src/JSInterop/Microsoft.JSInterop/src/JSRuntime.cs:line 111
   at System.Threading.CancellationToken.<>c.<Register>b__12_0(Object obj)
   at System.Threading.CancellationTokenSource.Invoke(Delegate d, Object state, CancellationTokenSource source)
   at System.Threading.CancellationTokenSource.CallbackNode.<>c.<ExecuteCallback>b__9_0(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.CancellationTokenSource.CallbackNode.ExecuteCallback()
   at System.Threading.CancellationTokenSource.ExecuteCallbackHandlers(Boolean throwOnFirstException)
   at System.Threading.CancellationTokenSource.NotifyCancellation(Boolean throwOnFirstException)
   at System.Threading.CancellationTokenSource.Cancel(Boolean throwOnFirstException)
   at System.Threading.CancellationTokenSource.Cancel()
   at Microsoft.AspNetCore.Components.Forms.BrowserFileStream.Dispose(Boolean disposing) in /workspaces/aspnetcore/src/Components/Web/src/Forms/InputFile/BrowserFileStream.cs:line 119
   at System.IO.Stream.Close()
   at System.IO.Stream.Dispose()
   at System.Net.Http.StreamContent.Dispose(Boolean disposing)
   at System.Net.Http.HttpContent.Dispose()
   at System.Net.Http.MultipartContent.Dispose(Boolean disposing)
   at System.Net.Http.HttpContent.Dispose()
   at BasicTestApp.FormsTest.InputFileComponent.OnInputFileChange(InputFileChangeEventArgs e) in /workspaces/aspnetcore/src/Components/test/testassets/BasicTestApp/FormsTest/InputFileComponent.razor:line 119
   at System.Reflection.RuntimeMethodInfo.InternalInvoke(RuntimeMethodInfo , Object , Span`1& , Exception& )
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at System.Delegate.DynamicInvokeImpl(Object[] args)
   at System.MulticastDelegate.DynamicInvokeImpl(Object[] args)
   at System.Delegate.DynamicInvoke(Object[] args)
   at Microsoft.AspNetCore.Components.EventCallbackWorkItem.InvokeAsync[Object](MulticastDelegate delegate, Object arg) in /workspaces/aspnetcore/src/Components/Components/src/EventCallbackWorkItem.cs:line 66
   at Microsoft.AspNetCore.Components.EventCallbackWorkItem.InvokeAsync(Object arg) in /workspaces/aspnetcore/src/Components/Components/src/EventCallbackWorkItem.cs:line 38
   at Microsoft.AspNetCore.Components.ComponentBase.Microsoft.AspNetCore.Components.IHandleEvent.HandleEventAsync(EventCallbackWorkItem callback, Object arg) in /workspaces/aspnetcore/src/Components/Components/src/ComponentBase.cs:line 303
   at Microsoft.AspNetCore.Components.EventCallback`1[[Microsoft.AspNetCore.Components.Forms.InputFileChangeEventArgs, Microsoft.AspNetCore.Components.Web, Version=7.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]].InvokeAsync(InputFileChangeEventArgs arg) in /workspaces/aspnetcore/src/Components/Components/src/EventCallbackOfT.cs:line 57
   at Microsoft.AspNetCore.Components.Forms.InputFile.Microsoft.AspNetCore.Components.Forms.IInputFileJsCallbacks.NotifyChange(BrowserFile[] files) in /workspaces/aspnetcore/src/Components/Web/src/Forms/InputFile.cs:line 110
   at Microsoft.AspNetCore.Components.Forms.InputFileJsCallbacksRelay.NotifyChange(BrowserFile[] files) in /workspaces/aspnetcore/src/Components/Web/src/Forms/InputFile/InputFileJsCallbacksRelay.cs:line 27
   at System.Reflection.RuntimeMethodInfo.InternalInvoke(RuntimeMethodInfo , Object , Span`1& , Exception& )
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at Microsoft.JSInterop.Infrastructure.DotNetDispatcher.InvokeSynchronously(JSRuntime jsRuntime, DotNetInvocationInfo& callInfo, IDotNetObjectReference objectReference, String argsJson) in /workspaces/aspnetcore/src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:line 172
   at Microsoft.JSInterop.Infrastructure.DotNetDispatcher.BeginInvokeDotNet(JSRuntime jsRuntime, DotNetInvocationInfo invocationInfo, String argsJson) in /workspaces/aspnetcore/src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:line 92
   at Microsoft.AspNetCore.Components.WebAssembly.Services.DefaultWebAssemblyJSRuntime.<>c.<BeginInvokeDotNet>b__8_0(ValueTuple`2 state) in /workspaces/aspnetcore/src/Components/WebAssembly/WebAssembly/src/Services/DefaultWebAssemblyJSRuntime.cs:line 79
   at Microsoft.AspNetCore.Components.WebAssembly.Hosting.WebAssemblyCallQueue.Schedule[ValueTuple`2](ValueTuple`2 state, Action`1 callback) in /workspaces/aspnetcore/src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyCallQueue.cs:line 58
   at Microsoft.AspNetCore.Components.WebAssembly.Services.DefaultWebAssemblyJSRuntime.BeginInvokeDotNet(String callId, String assemblyNameOrDotNetObjectId, String methodIdentifier, String argsJson) in /workspaces/aspnetcore/src/Components/WebAssembly/WebAssembly/src/Services/DefaultWebAssemblyJSRuntime.cs:line 75
   at Microsoft.JSInterop.JSRuntime.EndInvokeJS(Int64 taskId, Boolean succeeded, Utf8JsonReader& jsonReader) in /workspaces/aspnetcore/src/JSInterop/Microsoft.JSInterop/src/JSRuntime.cs:line 238
   at Microsoft.JSInterop.Infrastructure.DotNetDispatcher.EndInvokeJS(JSRuntime jsRuntime, String arguments) in /workspaces/aspnetcore/src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:line 309
   at Microsoft.AspNetCore.Components.WebAssembly.Services.DefaultWebAssemblyJSRuntime.<>c.<EndInvokeJS>b__7_0(String argsJson) in /workspaces/aspnetcore/src/Components/WebAssembly/WebAssembly/src/Services/DefaultWebAssemblyJSRuntime.cs:line 51
   at Microsoft.AspNetCore.Components.WebAssembly.Hosting.WebAssemblyCallQueue.Schedule[String](String state, Action`1 callback) in /workspaces/aspnetcore/src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyCallQueue.cs:line 58
   at Microsoft.AspNetCore.Components.WebAssembly.Services.DefaultWebAssemblyJSRuntime.EndInvokeJS(String argsJson) in /workspaces/aspnetcore/src/Components/WebAssembly/WebAssembly/src/Services/DefaultWebAssemblyJSRuntime.cs:line 53
    at _convert_exception_for_method_call (dotnet.7.0.0-alpha.1.21618.4.hp2ca9myzh.js:3)
    at _handle_exception_for_call (dotnet.7.0.0-alpha.1.21618.4.hp2ca9myzh.js:3)
    at _Microsoft_AspNetCore_Components_WebAssembly__Microsoft_AspNetCore_Components_WebAssembly_Services_DefaultWebAssemblyJSRuntime_EndInvokeJS (_Microsoft_AspNetCore_Components_WebAssembly__Microsoft_AspNetCore_Components_WebAssembly_Services_DefaultWebAssemblyJSRuntime_EndInvokeJS:24)
    at Object.endInvokeJSFromDotNet (MonoPlatform.ts:538)
    at Microsoft.JSInterop.js:303
```

Stack trace with args:

```
       Uncaught (in promise) Error: System.Text.Json.JsonException: Invalid JSON [5,true,{"__jsStreamReferenceLength":198598,"__jsObjectId":1}], Token: PropertyName, Position: 0
   at Microsoft.JSInterop.Infrastructure.DotNetDispatcher.EndInvokeJS(JSRuntime jsRuntime, String arguments) in /workspaces/aspnetcore/src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:line 313
   at Microsoft.AspNetCore.Components.WebAssembly.Services.DefaultWebAssemblyJSRuntime.<>c.<EndInvokeJS>b__7_0(String argsJson) in /workspaces/aspnetcore/src/Components/WebAssembly/WebAssembly/src/Services/DefaultWebAssemblyJSRuntime.cs:line 51
   at Microsoft.AspNetCore.Components.WebAssembly.Hosting.WebAssemblyCallQueue.Schedule[String](String state, Action`1 callback) in /workspaces/aspnetcore/src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyCallQueue.cs:line 58
   at Microsoft.AspNetCore.Components.WebAssembly.Services.DefaultWebAssemblyJSRuntime.EndInvokeJS(String argsJson) in /workspaces/aspnetcore/src/Components/WebAssembly/WebAssembly/src/Services/DefaultWebAssemblyJSRuntime.cs:line 53
    at _convert_exception_for_method_call (dotnet.7.0.0-alpha.1.21618.4.hp2ca9myzh.js:3)
    at _handle_exception_for_call (dotnet.7.0.0-alpha.1.21618.4.hp2ca9myzh.js:3)
    at _Microsoft_AspNetCore_Components_WebAssembly__Microsoft_AspNetCore_Components_WebAssembly_Services_DefaultWebAssemblyJSRuntime_EndInvokeJS (_Microsoft_AspNetCore_Components_WebAssembly__Microsoft_AspNetCore_Components_WebAssembly_Services_DefaultWebAssemblyJSRuntime_EndInvokeJS:24)
    at Object.endInvokeJSFromDotNet (MonoPlatform.ts:538)
    at Microsoft.JSInterop.js:303
```

</details>

**Comment by Grizzlly** [comment]
Just to clarify: The workaround is to use `StringContent` with `application/octet-stream`?

Also, I think you should reconsider and update #39075 as I think it affects everyone that implements file upload.

**Comment by TanayParikh** [comment]
> Just to clarify: The workaround is to use `StringContent` with `application/octet-stream`?

Yes. The reason why that works isn't because of `StringContent` but rather because `await fileContent.ReadAsStringAsync()` ensures we await the streaming initialization. 

`content: new StringContent(await fileContent.ReadAsStringAsync(), Encoding.UTF8, "application/octet-stream")`

> Also, I think you should reconsider and update #39075 as I think it affects everyone that implements file upload.

Yes we'll be discussing potentially bringing this into a future patch release in the new year once team members return from seasonal holidays. Fortunately it won't affect __everyone__ as those who do the actual upload end up reading the stream as well per the sample:

```csharp
        if (upload)
        {
            var response = await Http.PostAsync("/Filesave", content);

            var newUploadResults = await response.Content
                .ReadFromJsonAsync<IList<UploadResult>>();

            if (newUploadResults is not null)
            {
                uploadResults = uploadResults.Concat(newUploadResults).ToList();
            }
        }
```

This is present in the official docs (but not in the sample you provided). This is likely also the explanation for why you and I were seeing this error reproduce, however Luke was not (as Luke was likely using the `FileUpload2` component exactly as documented with the upload section).

**Comment by RobertBouillon** [comment]
Quick question - will the fix to this issue restore functionality the original use-cases?

Here's a succinct workaround. In my case, I'm reading bytes, not a string. I was buffering the data to a `MemoryStream` to be consumed by classes that take a `Stream` as a parameter. 

TLDR; Looks like you just have wrap the `Stream` returned by `IBrowserFile` with an `HttpContent` class (`StringContent`, `ByteArrayContent`, etc), and then async read from there.

Broken in .NET 6
```csharp
using var ms = new MemoryStream((int)file.Size);
using var x = file.OpenReadStream(MAX_FILE_SIZE);
await x.CopyToAsync(ms);
```
Equivalent workaround
```csharp
using var fileContent = new StreamContent(file.OpenReadStream(MAX_FILE_SIZE));
using var ms = new MemoryStream(await fileContent.ReadAsByteArrayAsync());
```

**Comment by TanayParikh** [comment]
> Quick question - will the fix to this issue restore functionality the original use-cases?

The example as [documented](https://docs.microsoft.com/en-us/aspnet/core/blazor/file-uploads?view=aspnetcore-6.0&pivots=webassembly#upload-component) works already (the issue is if you don't actually read from the stream on which you've called `OpenReadStream` and immediately cancel). This was fixed by https://github.com/dotnet/aspnetcore/pull/39060 and _may_ be backported to 6.0 via https://github.com/dotnet/aspnetcore/pull/39075.

> Broken in .NET 6

That example shouldn't be broken as we're actually reading the stream  via `CopyToAsync`. If you're facing this issue, could you please provide a minimal public github repro project so I may take a look. You shouldn't have to wrap the stream in a `StreamContent` like you do in your example.

**Comment by TanayParikh** [comment]
Hmm this repro shouldn't require MVC at all. You should just be able to do something in a clean wasm app (`dotnet new blazorwasm`).

**Comment by Grizzlly** [comment]
I guess one extra possible workaround would be to keep the browser files in a list and only add them to the DataContent before sending the request?

**Comment by TanayParikh** [comment]
> keep the browser files in a list and only add them to the DataContent before sending the request?

The issue is **not** with the DataContent, but rather calling `file.OpenReadStream` and then immediately cancelling the request (by not doing anything with the stream / letting the function terminate). If you don't call `file.OpenReadStream` you shouldn't have this issue.

**Comment by Grizzlly** [comment]
Yeah, I think I understand things more clearly now, thanks!

**Comment by TanayParikh** [comment]
https://github.com/dotnet/aspnetcore/pull/39075 is merged and the patch will be available in 6.0.2. Thanks all.
